# 1. The basetable timeline

In [1]:
import pandas as pd
import datetime

In [2]:
def min_max(column):
    return pd.Series(index=['min','max'], data=[column.min(), column.max()])

In [3]:
gifts = pd.read_csv('data2/gifts.csv')
gifts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 4 columns):
Unnamed: 0    150000 non-null int64
id            150000 non-null int64
date          150000 non-null object
amount        150000 non-null float64
dtypes: float64(1), int64(2), object(1)
memory usage: 4.6+ MB


In [4]:
gifts.head(2)

Unnamed: 0  id        date  amount
0           0   1  2015-10-16    75.0
1           1   1  2014-02-11   111.0

In [5]:
gifts['date'] = pd.to_datetime(gifts['date'])
gifts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 4 columns):
Unnamed: 0    150000 non-null int64
id            150000 non-null int64
date          150000 non-null datetime64[ns]
amount        150000 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(2)
memory usage: 4.6 MB


In [6]:
gifts.head(2)

Unnamed: 0  id       date  amount
0           0   1 2015-10-16    75.0
1           1   1 2014-02-11   111.0

In [7]:
gifts.apply(min_max)

Unnamed: 0     id       date  amount
min           0      1 2010-06-01     3.0
max      149999  27402 2017-11-30   195.0

In [8]:
gifts.describe()

Unnamed: 0             id         amount
count  150000.000000  150000.000000  150000.000000
mean    74999.500000   12988.146267      92.384007
std     43301.414527    8131.261153      22.977188
min         0.000000       1.000000       3.000000
25%     37499.750000    5461.750000      77.000000
50%     74999.500000   12779.000000      92.000000
75%    112499.250000   20080.000000     108.000000
max    149999.000000   27402.000000     195.000000

In [9]:
gifts.drop(['Unnamed: 0'], axis=1, inplace=True)
gifts.head()

id       date  amount
0   1 2015-10-16    75.0
1   1 2014-02-11   111.0
2   1 2012-03-28    93.0
3   1 2013-12-13   113.0
4   1 2012-01-10    93.0

In [10]:
start_target = datetime.datetime(year=2018, month=5, day=1)
end_target = datetime.datetime(year=2018, month=8, day=1)

start_target, end_target

(datetime.datetime(2018, 5, 1, 0, 0), datetime.datetime(2018, 8, 1, 0, 0))

In [11]:
gifts_target = gifts[(gifts['date']>=start_target) & (gifts['date']<end_target)]
gifts_target.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 3 columns):
id        0 non-null int64
date      0 non-null datetime64[ns]
amount    0 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 0.0 bytes


In [12]:
gifts_pred_variables = gifts[(gifts['date']<start_target)]
gifts_pred_variables.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 150000 entries, 0 to 149999
Data columns (total 3 columns):
id        150000 non-null int64
date      150000 non-null datetime64[ns]
amount    150000 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 4.6 MB


In [13]:
gifts_pred_variables.head(2)

id       date  amount
0   1 2015-10-16    75.0
1   1 2014-02-11   111.0

In [14]:
gifts_pred_variables.apply(min_max)

id       date  amount
min      1 2010-06-01     3.0
max  27402 2017-11-30   195.0

# 1.1. The population

In [15]:
donation_2016 = gifts[gifts['date'].dt.year==2016]
donation_2016.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16514 entries, 13 to 149991
Data columns (total 3 columns):
id        16514 non-null int64
date      16514 non-null datetime64[ns]
amount    16514 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 516.1 KB


In [16]:
donors_include = set(donation_2016['id'])
len(donors_include)

12062

In [17]:
donation_2017 = gifts[(gifts['date'].dt.year==2017) & (gifts['date'].dt.month<5)]
donation_2017.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2650 entries, 48 to 149921
Data columns (total 3 columns):
id        2650 non-null int64
date      2650 non-null datetime64[ns]
amount    2650 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 82.8 KB


In [18]:
donors_exclude = set(donation_2017['id'])
len(donors_exclude)

2305

In [19]:
population = donors_include.difference(donors_exclude)
len(population)

10934

Population is the list of people who unsubscribe in 2017. Since you have **12,062** ids in 2016 and **2,305** ids in 2017

# 1.2. The target

In [20]:
basetable = pd.read_csv('data2/basetable.csv')
basetable.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9506 entries, 0 to 9505
Data columns (total 1 columns):
donor_ID    9506 non-null int64
dtypes: int64(1)
memory usage: 74.3 KB


In [21]:
basetable.head()

donor_ID
0     32768
1     98305
2     65537
3     65545
4     32780

In [22]:
basetable.describe()

donor_ID
count   9506.000000
mean   50348.401851
std    28981.438932
min       29.000000
25%    25025.250000
50%    50278.500000
75%    75788.000000
max    99987.000000

In [23]:
basetable2 = pd.read_csv('data2/basetable_ex_2_13.csv')
basetable2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 6 columns):
Unnamed: 0                 100000 non-null int64
target                     100000 non-null int64
age                        100000 non-null int64
gender_F                   100000 non-null int64
donations_2017             100000 non-null int64
donations_2017_min_2016    100000 non-null int64
dtypes: int64(6)
memory usage: 4.6 MB


In [24]:
basetable2.head()

Unnamed: 0  target  age  gender_F  donations_2017  donations_2017_min_2016
0           0       1   42         0              10                        7
1           1       1   73         0               4                        1
2           2       1   37         0               6                        3
3           3       1   75         0               2                       -4
4           4       1   32         1               9                        5

In [25]:
basetable2.apply(min_max)

Unnamed: 0  target  age  gender_F  donations_2017  \
min           0       0   19         0               0   
max       99999       1  110         1              10   

     donations_2017_min_2016  
min                      -11  
max                       10

In [26]:
basetable3 = pd.read_csv('data2/basetable_interactions.csv')
basetable3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 5 columns):
id                10000 non-null int64
target            10000 non-null int64
country_Spain     10000 non-null int64
country_France    10000 non-null int64
age               10000 non-null float64
dtypes: float64(1), int64(4)
memory usage: 390.7 KB


In [27]:
basetable3.head()

id  target  country_Spain  country_France   age
0  1416       0              1               0  25.0
1  1046       0              1               0  44.0
2  8912       0              0               1  59.0
3  3531       0              1               0  57.0
4  7009       0              0               1  43.0

In [28]:
basetable3.apply(min_max)

id  target  country_Spain  country_France    age
min      1       0              0               0  -48.0
max  10000       1              1               1  128.0

In [29]:
living_places = pd.read_csv('data2/living_places.csv')
living_places.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28806 entries, 0 to 28805
Data columns (total 4 columns):
donor_ID        28806 non-null int64
start_date      28806 non-null object
end_date        28806 non-null object
living_place    28806 non-null object
dtypes: int64(1), object(3)
memory usage: 900.3+ KB


In [30]:
living_places.iloc[:, 1:3] = living_places.iloc[:, 1:3].apply(pd.to_datetime, errors='coerce')
living_places.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28806 entries, 0 to 28805
Data columns (total 4 columns):
donor_ID        28806 non-null int64
start_date      28806 non-null datetime64[ns]
end_date        28806 non-null datetime64[ns]
living_place    28806 non-null object
dtypes: datetime64[ns](2), int64(1), object(1)
memory usage: 900.3+ KB


In [31]:
living_places.head()

donor_ID start_date   end_date living_place
0     32768 1989-03-25 2000-07-06        India
1     32768 2000-07-06 2099-01-01           UK
2     98305 1954-04-06 1958-08-12          USA
3     98305 1958-08-12 1972-10-04           UK
4     98305 1972-10-04 1988-04-18           UK

In [32]:
living_places.apply(min_max)

donor_ID start_date   end_date living_place
min        29 1950-01-07 1950-05-13      Belgium
max     99987 2017-12-14 2099-01-01          USA

In [33]:
basetable['target'] = pd.Series([1 if donor_id in population else 0 for donor_id in basetable['donor_ID']])
basetable.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9506 entries, 0 to 9505
Data columns (total 2 columns):
donor_ID    9506 non-null int64
target      9506 non-null int64
dtypes: int64(2)
memory usage: 148.6 KB


In [34]:
basetable.head()

donor_ID  target
0     32768       0
1     98305       0
2     65537       0
3     65545       0
4     32780       0

In [35]:
basetable['target'].value_counts()

0    8462
1    1044
Name: target, dtype: int64

In [36]:
'''Target period'''
start_target = datetime.datetime(year=2017, month=1, day=1)
end_target = datetime.datetime(year=2018, month=1, day=1)

start_target, end_target

(datetime.datetime(2017, 1, 1, 0, 0), datetime.datetime(2018, 1, 1, 0, 0))

In [37]:
'''Target period donation'''
gifts_target = gifts[(gifts['date']>=start_target) & (gifts['date']<end_target)]
gifts_target.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4955 entries, 48 to 149966
Data columns (total 3 columns):
id        4955 non-null int64
date      4955 non-null datetime64[ns]
amount    4955 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 154.8 KB


In [38]:
gifts_target.head(2)

id       date  amount
48   8 2017-02-13   111.0
64  11 2017-02-24    96.0

In [39]:
'''Group and sum donations by donor'''
gifts_target_byid = gifts_target.groupby('id')['amount'].sum().reset_index()
gifts_target_byid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3843 entries, 0 to 3842
Data columns (total 2 columns):
id        3843 non-null int64
amount    3843 non-null float64
dtypes: float64(1), int64(1)
memory usage: 60.1 KB


In [40]:
gifts_target_byid.head(2)

id  amount
0   8   111.0
1  11    96.0

In [41]:
'''Derive targets and add to basetable'''
targets = list(gifts_target_byid['id'][gifts_target_byid['amount']>500])
targets

[222, 1321, 4508]

In [42]:
# basetable['target'] = pd.Series([1 if donor_id in gifts_target_byid['id'].values.tolist() else 0 for donor_id in basetable['donor_ID']])
basetable['target'] = pd.Series([1 if donor_id in targets else 0 for donor_id in basetable['donor_ID']])

basetable.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9506 entries, 0 to 9505
Data columns (total 2 columns):
donor_ID    9506 non-null int64
target      9506 non-null int64
dtypes: int64(2)
memory usage: 148.6 KB


In [43]:
basetable.head()

donor_ID  target
0     32768       0
1     98305       0
2     65537       0
3     65545       0
4     32780       0

In [44]:
basetable['target'].value_counts()

0    9506
Name: target, dtype: int64

# 2. Adding predictive variables

In [45]:
reference_date = datetime.datetime(2018,4,1)
reference_date

datetime.datetime(2018, 4, 1, 0, 0)

In [46]:
living_places.head(2)

donor_ID start_date   end_date living_place
0     32768 1989-03-25 2000-07-06        India
1     32768 2000-07-06 2099-01-01           UK

In [47]:
living_places['active_period'] = living_places['end_date'] - living_places['start_date']
living_places.head(2)

donor_ID start_date   end_date living_place active_period
0     32768 1989-03-25 2000-07-06        India     4121 days
1     32768 2000-07-06 2099-01-01           UK    35973 days

In [48]:
living_places['lifetime'] = reference_date - living_places['start_date']
living_places.head(2)

donor_ID start_date   end_date living_place active_period   lifetime
0     32768 1989-03-25 2000-07-06        India     4121 days 10599 days
1     32768 2000-07-06 2099-01-01           UK    35973 days  6478 days

In [49]:
living_places_reference_date = living_places[(living_places['start_date']<=reference_date) & (living_places['end_date']>reference_date)]
living_places_reference_date.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9506 entries, 1 to 28805
Data columns (total 6 columns):
donor_ID         9506 non-null int64
start_date       9506 non-null datetime64[ns]
end_date         9506 non-null datetime64[ns]
living_place     9506 non-null object
active_period    9506 non-null timedelta64[ns]
lifetime         9506 non-null timedelta64[ns]
dtypes: datetime64[ns](2), int64(1), object(1), timedelta64[ns](2)
memory usage: 519.9+ KB


In [50]:
living_places_reference_date.head(2)

donor_ID start_date   end_date living_place active_period  lifetime
1     32768 2000-07-06 2099-01-01           UK    35973 days 6478 days
6     98305 2007-11-30 2099-01-01        India    33270 days 3775 days

# 2.1. Adding aggregated variables

In [51]:
'''Start and end date of the aggregation method'''
start_date = datetime.datetime(2016,1,1)
end_date = datetime.datetime(2017,1,1)

start_date, end_date

(datetime.datetime(2016, 1, 1, 0, 0), datetime.datetime(2017, 1, 1, 0, 0))

In [52]:
'''Select gifts made in 2016'''
gifts_2016 = gifts[(gifts['date']>=start_date) & (gifts['date']<=end_date)]
gifts_2016.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16514 entries, 13 to 149991
Data columns (total 3 columns):
id        16514 non-null int64
date      16514 non-null datetime64[ns]
amount    16514 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 516.1 KB


In [53]:
gifts_2016.head(2)

id       date  amount
13   3 2016-03-31    90.0
24   3 2016-02-10   121.0

In [54]:
'''Sum of gifts per donor in 2016'''
gifts_2016_bydonor = gifts_2016.groupby(['id'])['amount'].sum().reset_index()
gifts_2016_bydonor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12062 entries, 0 to 12061
Data columns (total 2 columns):
id        12062 non-null int64
amount    12062 non-null float64
dtypes: float64(1), int64(1)
memory usage: 188.5 KB


In [55]:
gifts_2016_bydonor.columns = ['donor_ID','sum_2016']
gifts_2016_bydonor.head(2)

donor_ID  sum_2016
0         3     211.0
1         4     234.0

In [56]:
basetable.head(2)

donor_ID  target
0     32768       0
1     98305       0

In [57]:
'''Add sum of gifts to the basetable'''
basetable = pd.merge(basetable, gifts_2016_bydonor, how='left', on='donor_ID')
basetable.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9506 entries, 0 to 9505
Data columns (total 3 columns):
donor_ID    9506 non-null int64
target      9506 non-null int64
sum_2016    1139 non-null float64
dtypes: float64(1), int64(2)
memory usage: 297.1 KB


In [58]:
basetable.head(2)

donor_ID  target  sum_2016
0     32768       0       NaN
1     98305       0       NaN

In [59]:
'''Number of gifts per donor in 2016'''
gifts_2016_bydonor = gifts_2016.groupby(['id']).size().reset_index()
gifts_2016_bydonor.columns = ['donor_ID','count_2016']
gifts_2016_bydonor.head(2)

donor_ID  count_2016
0         3           2
1         4           2

In [60]:
'''Add sum of gifts to the basetable'''
basetable = pd.merge(basetable, gifts_2016_bydonor, how='left', on='donor_ID')
basetable.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9506 entries, 0 to 9505
Data columns (total 4 columns):
donor_ID      9506 non-null int64
target        9506 non-null int64
sum_2016      1139 non-null float64
count_2016    1139 non-null float64
dtypes: float64(2), int64(2)
memory usage: 371.3 KB


In [61]:
basetable.head(2)

donor_ID  target  sum_2016  count_2016
0     32768       0       NaN         NaN
1     98305       0       NaN         NaN

# 2.2. Adding evolutions

In [62]:
start_2017 = datetime.datetime(2017,1,1)
start_2016 = datetime.datetime(2016,1,1)
start_2015 = datetime.datetime(2015,1,1)

start_2015, start_2016, start_2017

(datetime.datetime(2015, 1, 1, 0, 0),
 datetime.datetime(2016, 1, 1, 0, 0),
 datetime.datetime(2017, 1, 1, 0, 0))

In [63]:
gifts_2016 = gifts[(gifts['date']<start_2017) & (gifts['date']>=start_2016)]
gifts_2016.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16514 entries, 13 to 149991
Data columns (total 3 columns):
id        16514 non-null int64
date      16514 non-null datetime64[ns]
amount    16514 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 516.1 KB


In [64]:
gifts_2016.head(2)

id       date  amount
13   3 2016-03-31    90.0
24   3 2016-02-10   121.0

In [65]:
gifts_2015_and_2016 = gifts[(gifts['date']<start_2017) & (gifts['date']>=start_2015)]
gifts_2015_and_2016.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40126 entries, 0 to 149996
Data columns (total 3 columns):
id        40126 non-null int64
date      40126 non-null datetime64[ns]
amount    40126 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(1)
memory usage: 1.2 MB


In [66]:
gifts_2015_and_2016.head(2)

id       date  amount
0   1 2015-10-16    75.0
5   1 2015-04-22    85.0

In [67]:
number_gifts_2016 = gifts_2016.groupby(['id'])['amount'].size().reset_index()
number_gifts_2016.columns = ['donor_ID','number_gifts_2016']
number_gifts_2016.head(2)

donor_ID  number_gifts_2016
0         3                  2
1         4                  2

In [68]:
number_gifts_2015_and_2016 = gifts_2015_and_2016.groupby(['id'])['amount'].size().reset_index()
number_gifts_2015_and_2016.columns = ['donor_ID','number_gifts_2015_and_2016']
number_gifts_2015_and_2016.head(2)

donor_ID  number_gifts_2015_and_2016
0         1                           2
1         2                           1

In [69]:
basetable = pd.merge(basetable, number_gifts_2016, on='donor_ID', how='left')
basetable = pd.merge(basetable, number_gifts_2015_and_2016, on='donor_ID', how='left')
basetable.head(2)

donor_ID  target  sum_2016  count_2016  number_gifts_2016  \
0     32768       0       NaN         NaN                NaN   
1     98305       0       NaN         NaN                NaN   

   number_gifts_2015_and_2016  
0                         NaN  
1                         NaN

In [70]:
basetable.describe()

donor_ID  target     sum_2016   count_2016  number_gifts_2016  \
count   9506.000000  9506.0  1139.000000  1139.000000        1139.000000   
mean   50348.401851     0.0   125.320457     1.351185           1.351185   
std    28981.438932     0.0    67.440243     0.649135           0.649135   
min       29.000000     0.0    28.000000     1.000000           1.000000   
25%    25025.250000     0.0    83.000000     1.000000           1.000000   
50%    50278.500000     0.0   103.000000     1.000000           1.000000   
75%    75788.000000     0.0   143.500000     2.000000           2.000000   
max    99987.000000     0.0   601.000000     5.000000           5.000000   

       number_gifts_2015_and_2016  
count                 2098.000000  
mean                     1.837464  
std                      1.076824  
min                      1.000000  
25%                      1.000000  
50%                      1.000000  
75%                      2.000000  
max                      9.000000

In [71]:
'''Calculate ratio of last month and last year average'''
basetable['ratio_2015_to_2015_and_2016'] = basetable['number_gifts_2016'] / basetable['number_gifts_2015_and_2016']
basetable.describe()

donor_ID  target     sum_2016   count_2016  number_gifts_2016  \
count   9506.000000  9506.0  1139.000000  1139.000000        1139.000000   
mean   50348.401851     0.0   125.320457     1.351185           1.351185   
std    28981.438932     0.0    67.440243     0.649135           0.649135   
min       29.000000     0.0    28.000000     1.000000           1.000000   
25%    25025.250000     0.0    83.000000     1.000000           1.000000   
50%    50278.500000     0.0   103.000000     1.000000           1.000000   
75%    75788.000000     0.0   143.500000     2.000000           2.000000   
max    99987.000000     0.0   601.000000     5.000000           5.000000   

       number_gifts_2015_and_2016  ratio_2015_to_2015_and_2016  
count                 2098.000000                  1139.000000  
mean                     1.837464                     0.736724  
std                      1.076824                     0.274596  
min                      1.000000                     0.166667  
25%                      1.000000                     0.500000  
50%                      1.000000                     0.750000  
75%                      2.000000                     1.000000  
max                      9.000000                     1.000000

# 2.3. Using evolution variables

In [72]:
from sklearn import linear_model

In [73]:
variables = ['gender','age','donation_last_year','ratio_month_year']

Decided to pause this here because the data given and the notes are too different